src_Core/CPU/CPU.bsv

```diff
@@ -831,8 +905,23 @@ module mkCPU (CPU_IFC);
 	 // CSRRx is done off-pipe
 	 csr_regfile.csr_minstret_incr;
 	 fa_emit_instr_trace (minstret, stage2.out.data_to_stage3.pcc, stage2.out.data_to_stage3.instr, rg_cur_priv);
+	 
+`ifdef PERFORMANCE_MONITORING
+	 coreEvents.evt_SC_SUCCESS = stage2.perf.sc_success;
+`endif
       end
 
+`ifdef PERFORMANCE_MONITORING
+      coreEvents.evt_LD_WAIT = stage2.perf.ld_wait;
+      coreEvents.evt_ST_WAIT = stage2.perf.st_wait;
+
+      coreEvents.evt_WAIT_F = stageF_full;
+      coreEvents.evt_WAIT_D = stageD_full && (stageF.out.ostatus == OSTATUS_PIPE);
+      coreEvents.evt_WAIT_1 = stage1_full && (stageD.out.ostatus == OSTATUS_PIPE);
+      coreEvents.evt_WAIT_2 = stage2_full && (stage1.out.ostatus == OSTATUS_PIPE);
+      coreEvents.evt_WAIT_3 = stage3_full && (stage2.out.ostatus == OSTATUS_PIPE);
```

Comment:
I don't think that's useful though? There are four cases:
1. Stage N empty/busy
2. Stage N done but stage N+1 busy (/ trapping)
3. Stage N done and stage N+1 done but stage N+1 won't be deq'ed
4. Stage N done and stage N+1 done and stage N+1 will be deq'ed

Currently you're counting 2+3+4, but 4 is more like 1 than 2+3 (and, for "nice" code, a very common case) since as far as stage N is concerned, the instruction is moving down the pipeline this cycle (it's a pipeline, so stage N+1 can be full, deq'ed and then enq'ed with stage N's instruction in that order all in the same cycle).

Reply:
Currently I think using `stageN_full` is a mistake, it's value varies within the rule. I agree that currently 2+3+4 are counted and that we only want 2+3.
Maybe we only want to count 2 (the only place I can see 3 happening is where stage 1 won't be deq'ed due to a redirect and stage F being busy - is this the 'fault' of stage 1?).
To count 2 only I think one could use:
`(stageN.out.ostatus == OSTATUS_PIPE) && (stageN+1.out.ostatus != OSTATUS_PIPE) && (stageN+1.out.ostatus != OSTATUS_EMPTY)`
Using your suggestion from the first comment one could count 2+3 (but start with `stageN.out.ostatus == OSTATUS_PIPE` rather than `stageN_full` which would also count stage N busy/nonpipe)